File: linkis-computation-governance/linkis-client/linkis-computation-client/src/main/scala/org/apache/linkis/computation/client/interactive/InteractiveJob.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.computation.client.interactive

import org.apache.linkis.computation.client.{LinkisJobBuilder, ResultSetIterable}
import org.apache.linkis.computation.client.job.{
  StorableExistingLinkisJob,
  StorableLinkisJob,
  StorableSubmittableLinkisJob
}
import org.apache.linkis.ujes.client.UJESClient
import org.apache.linkis.ujes.client.request.JobSubmitAction
import org.apache.linkis.ujes.client.response.{JobInfoResult, JobLogResult, JobProgressResult}

import java.util

import scala.collection.mutable.ArrayBuffer

trait InteractiveJob extends StorableLinkisJob {

  private var resultSetList: Array[String] = _
  private val logListeners: ArrayBuffer[LogListener] = new ArrayBuffer[LogListener]()

  private val progressListeners: ArrayBuffer[ProgressListener] =
    new ArrayBuffer[ProgressListener]()

  protected var lastJobLogResult: JobLogResult = _
  protected var lastProgress: JobProgressResult = _

  def existResultSets: Boolean = wrapperId {
    resultSetList = getJobInfoResult.getResultSetList(ujesClient)
    resultSetList != null && resultSetList.nonEmpty
  }

  def getResultSetIterables: Array[ResultSetIterable] = {
    if (!existResultSets) return Array.empty
    resultSetList.map(
      new ResultSetIterable(ujesClient, getJobSubmitResult.getUser, _, getJobMetrics)
    )
  }

  def addLogListener(logListener: LogListener): Unit = {
    logListeners += logListener
    initJobDaemon()
  }

  def addProgressListener(progressListener: ProgressListener): Unit = {
    progressListeners += progressListener
    initJobDaemon()
  }

  override protected def getJobInfoResult: JobInfoResult = {
    lastJobLogResult =
      if (lastJobLogResult == null) ujesClient.log(getJobSubmitResult, 0, 1)
      else ujesClient.log(getJobSubmitResult, lastJobLogResult)
    logListeners.foreach(_.onLogUpdate(lastJobLogResult.getLog))
    val progress = ujesClient.progress(getJobSubmitResult)
    if (lastProgress == null || progress.getProgress > lastProgress.getProgress) {
      lastProgress = progress
      progressListeners.foreach(
        _.onProgressUpdate(lastProgress.getProgress, lastProgress.getProgressInfo)
      )
    }
    super.getJobInfoResult
  }

  def getLogListeners: ArrayBuffer[LogListener] = logListeners

  def getProgressListener: ArrayBuffer[ProgressListener] = progressListeners

}

class SubmittableInteractiveJob(ujesClient: UJESClient, jobSubmitAction: JobSubmitAction)
    extends StorableSubmittableLinkisJob(ujesClient, jobSubmitAction)
    with InteractiveJob {

  private var maxRetry = 0
  private var retryNumber = 0

  private var finishedJobInfoResult: JobInfoResult = _

  override protected def getJobInfoResult: JobInfoResult = {
    lastJobLogResult =
      if (lastJobLogResult == null) ujesClient.log(getJobSubmitResult, 0, 1)
      else ujesClient.log(getJobSubmitResult, lastJobLogResult)
    getLogListeners.foreach(_.onLogUpdate(lastJobLogResult.getLog))
    val progress = ujesClient.progress(getJobSubmitResult)
    if (lastProgress == null || progress.getProgress > lastProgress.getProgress) {
      lastProgress = progress
      getProgressListener.foreach(
        _.onProgressUpdate(lastProgress.getProgress, lastProgress.getProgressInfo)
      )
    }
    if (finishedJobInfoResult != null) return finishedJobInfoResult
    val startTime = System.currentTimeMillis
    val oldJobSubmitResult = getJobSubmitResult
    val jobInfoResult = wrapperId(ujesClient.getJobInfo(getJobSubmitResult))
    getJobMetrics.addClientGetJobInfoTime(System.currentTimeMillis - startTime)
    if (jobInfoResult.isCompleted) {
      if (jobInfoResult.isFailed && jobInfoResult.canRetry && getRetryNumber < getMaxRetry) {
        val retryMsg =
          s"Task ${oldJobSubmitResult.taskID} failed to run, retrying automatically, retry number: ${getRetryNumber}"
        logger.info(retryMsg)
        this.submit()
        addRetryNumber
        val logs = new util.ArrayList[String]()
        val retrySucceedMsg =
          s"The task has been retried, and the new task id is: ${getJobSubmitResult.taskID}"
        logs.add(retryMsg)
        logs.add(retrySucceedMsg)
        getLogListeners.foreach(_.onLogUpdate(logs))
        logger.info(retrySucceedMsg)
        return getJobInfoResult
      }
      getJobMetrics.setClientFinishedTime(System.currentTimeMillis)
      finishedJobInfoResult = jobInfoResult
      logger.info(s"Job-$getId is completed with status " + finishedJobInfoResult.getJobStatus)
      getJobListeners.foreach(_.onJobFinished(this))
    } else if (jobInfoResult.isRunning) getJobListeners.foreach(_.onJobRunning(this))
    jobInfoResult
  }

  def getMaxRetry: Int = maxRetry

  def setMaxRetry(maxRetry: Int): Unit = this.maxRetry = maxRetry

  def addRetryNumber: Unit = retryNumber = retryNumber + 1

  def getRetryNumber: Int = retryNumber

}

class ExistingInteractiveJob(ujesClient: UJESClient, execId: String, taskId: String, user: String)
    extends StorableExistingLinkisJob(ujesClient, execId, taskId, user)
    with InteractiveJob

object InteractiveJob {

  def builder(): InteractiveJobBuilder = new InteractiveJobBuilder

  /**
   * When use this method to create a InteractiveJob, ProgressListener and LogListener cannot be
   * used, because execID does not exist.
   *
   * @param taskId
   *   the id of InteractiveJob
   * @param user
   *   the execute user of InteractiveJob
   * @return
   */
  def build(taskId: String, user: String): InteractiveJob =
    new ExistingInteractiveJob(LinkisJobBuilder.getDefaultUJESClient, null, taskId, user)

  /**
   * Use this method to create a InteractiveJob, it is equivalent with [[builder]].
   *
   * @param execId
   *   the execId of InteractiveJob
   * @param taskId
   *   the taskId of InteractiveJob
   * @param user
   *   the execute user of InteractiveJob
   * @return
   */
  def build(execId: String, taskId: String, user: String): InteractiveJob =
    new ExistingInteractiveJob(LinkisJobBuilder.getDefaultUJESClient, null, taskId, user)

}

File: linkis-computation-governance/linkis-client/linkis-cli/src/main/java/org/apache/linkis/cli/application/interactor/job/interactive/InteractiveJob.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.linkis.cli.application.interactor.job.interactive;

import org.apache.linkis.cli.application.constants.CliConstants;
import org.apache.linkis.cli.application.constants.CliKeys;
import org.apache.linkis.cli.application.entity.context.CliCtx;
import org.apache.linkis.cli.application.entity.job.Job;
import org.apache.linkis.cli.application.entity.job.JobResult;
import org.apache.linkis.cli.application.exception.LinkisClientExecutionException;
import org.apache.linkis.cli.application.exception.LinkisClientRuntimeException;
import org.apache.linkis.cli.application.exception.error.CommonErrMsg;
import org.apache.linkis.cli.application.exception.error.ErrorLevel;
import org.apache.linkis.cli.application.interactor.job.common.JobKiller;
import org.apache.linkis.cli.application.interactor.job.common.LogRetriever;
import org.apache.linkis.cli.application.interactor.job.common.ResultRetriever;
import org.apache.linkis.cli.application.operator.OperManager;
import org.apache.linkis.cli.application.operator.ujes.LinkisJobOper;
import org.apache.linkis.cli.application.operator.ujes.LinkisOperResultAdapter;
import org.apache.linkis.cli.application.present.LogPresenter;
import org.apache.linkis.cli.application.present.ResultPresenter;
import org.apache.linkis.cli.application.utils.CliUtils;
import org.apache.linkis.cli.application.utils.LoggerManager;

import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.exception.ExceptionUtils;

import java.util.HashMap;
import java.util.Map;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class InteractiveJob implements Job {

  private static final Logger logger = LoggerFactory.getLogger(InteractiveJob.class);

  private CliCtx ctx;

  private Boolean isAsync = false;

  private LinkisJobOper oper;

  private InteractiveJobDesc desc;

  private String username;

  private String jobId;

  @Override
  public void build(CliCtx ctx) {
    this.ctx = ctx;
    this.isAsync =
        ctx.getVarAccess().getVarOrDefault(Boolean.class, CliKeys.LINKIS_CLIENT_ASYNC_OPT, false);
    this.desc = InteractiveJobDescBuilder.build(ctx);
    this.oper = (LinkisJobOper) OperManager.getNew(CliKeys.Linkis_OPER, ctx);
  }

  @Override
  public JobResult run() {

    // Indicator
    StringBuilder infoBuilder = new StringBuilder();
    infoBuilder.append("connecting to linkis gateway:").append(oper.getServerUrl());
    LoggerManager.getInformationLogger().info(infoBuilder.toString());
    infoBuilder.setLength(0);

    // Submit
    LinkisOperResultAdapter submitResult = oper.submit(desc);
    CliUtils.doSleepQuietly(CliConstants.JOB_QUERY_SLEEP_MILLS);

    // JobInfo
    LinkisOperResultAdapter jobInfoResult =
        oper.queryJobInfo(submitResult.getUser(), submitResult.getJobID());
    oper.queryJobStatus(
        submitResult.getUser(), submitResult.getJobID(), submitResult.getStrongerExecId());
    infoBuilder.setLength(0);
    infoBuilder
        .append("JobId:")
        .append(submitResult.getJobID())
        .append(System.lineSeparator())
        .append("TaskId:")
        .append(submitResult.getJobID())
        .append(System.lineSeparator())
        .append("ExecId:")
        .append(submitResult.getStrongerExecId());
    LoggerManager.getPlaintTextLogger().info(infoBuilder.toString());
    infoBuilder.setLength(0);

    // Submit success or not
    if (!jobInfoResult.getJobStatus().isJobSubmitted()) {
      return new InteractiveJobResult(false, "Failed to submit job", new HashMap<>());
    } else {
      // Output that job is submitted
      infoBuilder.append("Job is successfully submitted!").append(System.lineSeparator());
      LoggerManager.getInformationLogger().info(infoBuilder.toString());
      infoBuilder.setLength(0);
      username = submitResult.getUser();
      jobId = submitResult.getJobID();
    }

    // async job, return
    if (isAsync) {
      return new InteractiveJobResult(
          jobInfoResult.getJobStatus().isJobSubmitted(),
          "Async Submission Success",
          new HashMap<>());
    }

    CliUtils.doSleepQuietly(2000l);

    // get log while running
    LogRetriever logRetriever =
        new LogRetriever(
            jobInfoResult.getUser(),
            jobInfoResult.getJobID(),
            jobInfoResult.getStrongerExecId(),
            true,
            oper,
            new LogPresenter());
    // async because we need to query job status
    logRetriever.retrieveLogAsync();

    // wait complete
    jobInfoResult = waitJobComplete(submitResult.getUser(), submitResult.getJobID(), submitResult.getStrongerExecId());
    logRetriever.waitIncLogComplete();

    // get result-set
    String outputPath =
        ctx.getVarAccess().getVar(String.class, CliKeys.LINKIS_CLIENT_COMMON_OUTPUT_PATH);
    ResultPresenter presenter;
    if (StringUtils.isBlank(outputPath)) {
      presenter = new ResultPresenter();
    } else {
      presenter = new ResultPresenter(true, outputPath);
    }

    ResultRetriever resultRetriever =
        new ResultRetriever(
            jobInfoResult.getUser(),
            jobInfoResult.getJobID(),
            jobInfoResult.getStrongerExecId(),
            oper,
            presenter);

    JobResult result = getResult(jobInfoResult, resultRetriever);

    return result;
  }

  private JobResult getResult(
      LinkisOperResultAdapter jobInfoResult, ResultRetriever resultRetriever)
      throws LinkisClientRuntimeException {
    if (!jobInfoResult.getJobStatus().isJobSuccess()) {
      LoggerManager.getInformationLogger()
          .info(
              "Job status is not success but \'"
                  + jobInfoResult.getJobStatus()
                  + "\'. Will not try to retrieve any Result");
      Map<String, String> extraMap = new HashMap<>();
      if (jobInfoResult.getErrCode() != null) {
        extraMap.put("errorCode", String.valueOf(jobInfoResult.getErrCode()));
      }
      if (StringUtils.isNotBlank(jobInfoResult.getErrDesc())) {
        extraMap.put("errorDesc", jobInfoResult.getErrDesc());
      }
      return new InteractiveJobResult(false, "Execute Error!!!", extraMap);
    }
    InteractiveJobResult result =
        new InteractiveJobResult(true, "Execute Success!!!", new HashMap<>());
    try {
      resultRetriever.retrieveResultSync();
      result.setSuccess(true);
      result.setMessage("execute success!!!");
    } catch (LinkisClientExecutionException e) {
      if (e.getCode().equals("EXE0037")) {
        result.setSuccess(true);
        result.setMessage("execute success!!!");
        LoggerManager.getInformationLogger().warn(e.getMessage());
      } else {
        result.setSuccess(false);
        result.setMessage("execute failed!!!\n" + ExceptionUtils.getStackTrace(e));
      }
      resultRetriever.setResultFin(); // inform listener to stop
    } catch (Exception e) {
      result.setSuccess(false);
      result.setMessage("execute failed!!!\n" + ExceptionUtils.getStackTrace(e));
      resultRetriever.setResultFin(); // inform listener to stop
    }
    return result;
  }

  private LinkisOperResultAdapter waitJobComplete(String user, String jobId, String execId)
      throws LinkisClientRuntimeException {
    int retryCnt = 0;
    final int MAX_RETRY = 30;

    LinkisOperResultAdapter jobInfoResult = oper.queryJobInfo(user, jobId);
    oper.queryJobStatus(user, jobId, execId);

    while (!jobInfoResult.getJobStatus().isJobFinishedState()) {
      // query progress
      try {
        jobInfoResult = oper.queryJobInfo(user, jobId);
        oper.queryJobStatus(
            jobInfoResult.getUser(), jobInfoResult.getJobID(), execId);
      } catch (Exception e) {
        logger.warn("", e);
        retryCnt++;
        if (retryCnt >= MAX_RETRY) {
          throw new LinkisClientExecutionException(
              "EXE0013",
              ErrorLevel.ERROR,
              CommonErrMsg.ExecutionErr,
              "Cannot get jobStatus from server continuously for {0} seconds. Client aborted! Error message: \n",
              MAX_RETRY * 5 * CliConstants.JOB_QUERY_SLEEP_MILLS,
              e);
        }
        CliUtils.doSleepQuietly(
            5 * CliConstants.JOB_QUERY_SLEEP_MILLS); // maybe server problem. sleep
        // longer
        continue;
      }
      retryCnt = 0; // reset counter
      if (jobInfoResult.getJobStatus().isJobAbnormalStatus()) {
        throw new LinkisClientExecutionException(
            "EXE0006",
            ErrorLevel.ERROR,
            CommonErrMsg.ExecutionErr,
            "Job is in abnormal status: " + CliUtils.GSON.toJson(jobInfoResult));
      }
      CliUtils.doSleepQuietly(CliConstants.JOB_QUERY_SLEEP_MILLS);
    }
    return jobInfoResult;
  }

  @Override
  public void onDestroy() {
    if (StringUtils.isBlank(username) || StringUtils.isBlank(jobId)) {
      logger.warn("Failed to kill job username or jobId is blank");
      return;
    }
    if (isAsync) {
      return;
    }
    try {
      new JobKiller(oper).doKill(username, jobId);
    } catch (Exception e) {
      logger.error("Failed to kill job", e);
    }
  }
}
